sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-scheduler.ts

```diff
@@ -84,29 +87,81 @@ export abstract class BaseScheduler<T> implements OnApplicationBootstrap {
       await this.schedulerQueue.obliterate({ force: true });
       return;
     }
-    await this.deleteOldRepeatableJobs();
-    // Add the cron to the queue.
-    await this.schedulerQueue.add(await this.payload());
+    // Check if the job is paused to avoid creating new delayed jobs.
+    const isPaused = await this.schedulerQueue.isPaused();
+    if (isPaused) {
+      return;
+    }
+    // Acquire a lock based on the queue name to prevent multiple
+    // queue-consumers instances to be initialized at the same time
+    // and execute the checks concurrently.
+    await this.queueService.acquireQueueLock(
+      this.schedulerQueue.name as QueueNames,
+      async () => {
+        await this.ensureNextDelayedJobCreation();
+      },
+    );
   }
 
   /**
-   * Check if there is any old cron job  (i.e whenever there is a
-   * change in cron option, then a new job is created the old job
-   * will be still there in the queue) and delete it and add the
-   * new job to the queue.
-   * Note: If there is an old retrying job, it won't be deleted,
-   * as "getRepeatableJobs" will not fetch retrying jobs.
+   * Ensures a scheduled job will have a delayed job created with the
+   * next expected scheduled time based on the configured cron expression.
    */
-  private async deleteOldRepeatableJobs(): Promise<void> {
-    const getAllRepeatableJobs = await this.schedulerQueue.getRepeatableJobs();
-    const cronRepeatOption = await this.queueCronConfiguration();
-    getAllRepeatableJobs.forEach((job) => {
-      if (job.cron !== cronRepeatOption.cron) {
-        this.schedulerQueue.removeRepeatableByKey(job.key);
+  private async ensureNextDelayedJobCreation(): Promise<void> {
+    const delayedJobs = await this.schedulerQueue.getDelayed();
+    const nextExpectedScheduledJobDate =
+      await this.getNexSchedulerExecutionDate();
+    // Check if there is a delayed job with the expected scheduled time.
+    const expectedDelayedJob = delayedJobs.find((delayedJob) => {
+      const nextExecution = this.getDelayedJobExecutionDate(delayedJob);
+      return dayjs(nextExecution).isSame(nextExpectedScheduledJobDate);
+    });
+    // If the only delayed job is the expected one, no further verifications are needed.
+    if (expectedDelayedJob && delayedJobs.length === 1) {
+      return;
+    }
+    // Remove any non expected delayed job.
+    for (const delayedJob of delayedJobs) {
+      if (!expectedDelayedJob || delayedJob !== expectedDelayedJob) {
```

Comment:
Asking for my understanding, will comparing the job ids alone work ? `delayedJob.id !== expectedDelayedJob.id`

Reply:
The idea here is to compare the object reference, which will be pretty precise.
Is there a benefit using the id?